applications/OptimizationApplication/custom_utilities/control/sigmoidal_projection_utils.cpp

```diff
@@ -0,0 +1,290 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   license: OptimizationApplication/license.txt
+//
+//  Main author:     Reza Najian Asl
+//                   Suneth Warnakulasuriya
+//
+
+// System includes
+#include <sstream>
+#include <type_traits>
+
+// Project includes
+#include "expression/variable_expression_io.h"
+#include "expression/variable_expression_data_io.h"
+#include "expression/container_data_io.h"
+#include "includes/define.h"
+#include "includes/variables.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/reduction_utilities.h"
+#include "utilities/variable_utils.h"
+
+// Application includes
+#include "optimization_application_variables.h"
+
+// Include base h
+#include "sigmoidal_projection_utils.h"
+
+namespace Kratos
+{
+
+///@name Kratos Classes
+///@{
+
+bool HasVectorDuplicates(
+    std::vector<double> values){
+        std::sort(values.begin(), values.end());
+        return std::adjacent_find(values.begin(), values.end()) != values.end();
+};
+
+void CheckXYVectors(std::vector<double> xValues, std::vector<double> yValues){
+    if (xValues.size() != yValues.size())
+        KRATOS_ERROR << "SigmoidalProjectionUtils: xLimits and yLimits should have the same size.\n";
+
+    if (xValues.size() < 2)
+        KRATOS_ERROR << "SigmoidalProjectionUtils: xLimits and yLimits should have at least two entries.\n";
+
+    if (!std::is_sorted(xValues.begin(), xValues.end()))
+        KRATOS_ERROR << "SigmoidalProjectionUtils: xLimits should be sorted ascending.\n";
+
+    if (!std::is_sorted(yValues.begin(), yValues.end()))
+        KRATOS_ERROR << "SigmoidalProjectionUtils: yValues should be sorted ascending.\n";
+
+    if (HasVectorDuplicates(xValues))
+        KRATOS_ERROR << "SigmoidalProjectionUtils: xValues have duplications.\n";
+
+    if (HasVectorDuplicates(yValues))
+        KRATOS_ERROR << "SigmoidalProjectionUtils: yValues have duplications.\n";
+}
+
+double ProjectValueForward(
+        double xValue,
+        std::vector<double> xLimits,
+        std::vector<double> yLimits,
+        double beta,
+        int penalFac)
+{
+
+    size_t size = xLimits.size();
+    double x1 = xLimits[0], x2 = xLimits[1], y1 = yLimits[0], y2 = yLimits[1];
+
+    for (size_t i = 0; i < size - 1; ++i) {
+        if (xValue >= xLimits[i] && xValue <= xLimits[i + 1]) {
+            x1 = xLimits[i];
+            x2 = xLimits[i + 1];
+            y1 = yLimits[i];
+            y2 = yLimits[i + 1];
+            break;
+        }
+    }
+
+    if (xValue > xLimits[size - 1]) {
+        x1 = xLimits[size - 2];
+        x2 = xLimits[size - 1];
+        y1 = yLimits[size - 2];
+        y2 = yLimits[size - 1];
+    }
```

Comment:
Since `xLimits` and `yLimits` needs and checked to have to be in ascending order, this can be simplified to following with less `if` checks, less `reads` and less `writes`:
```suggestion
    size_t upper_index;
    for (upper_index = 0; upper_index < size; ++upper_index) {
        if (xValue < xLimits[upper_index]) {
            break;
        }
    }
    
    upper_index = std::clamp(upper_index, 1, size - 1);
       
    const double x1 = xLimits[upper_index - 1];
    const double x2 = xLimits[upper_index];
    const double y1 = yLimits[upper_index - 1];
    const double y2 = yLimits[upper_index];
```